### ISSUE TITLE
If a transaction will not succeed, could resubmitting it leak information?

### ISSUE THREAD

**Issue by daira** [zcash/zcash#679]
Suppose that I submit a transaction containing Pours, and then find that it won't be able to succeed, e.g. it depends on an output that is already spent. Could resubmitting a modified version of it (possibly with some of the same confidential input coins) leak information about me?

**Comment by ebfull** [comment]
I think the only thing it leaks is that both `Pour`s spend the same or some of the same coins, and I'm not sure if that's a bad thing or not.

Users should not be anchored to recent block treestates (as we've discussed before) because a reorg will permanently disqualify the `Pour` and require such a recalculation. If the recalculation is dangerous in some way, that is bad.

**Comment by defuse** [comment]
I was just thinking about this and decided to name it "reorg induced pour failure." If I use the Merkle root of the latest block to construct a pour of some coins I own, and then a reorg happens, the pour I just created won't be spendable on the longer branch. There's a privacy problem because an attacker might have seen me try to publish my pour to the shorter branch, and they'll notice when I retry on the longer branch since the serial numbers will be the same.

There's also a *usability* problem, which is firstly that pours are more likely to fail than cleartext transactions, since cleartext transactions don't need to be recreated when there's a reorg. Secondly, and more importantly, is if I'm trying to combine coins in a privacy-preserving way, and that'll take, say 10 blocks, then if I see a reorg any time during those 10 blocks the process will fail partway through.

**Comment by defuse** [comment]
It's a privacy problem because it tells the observer *that I saw a reorg* and the set of people who saw a reorg can be smaller than the set of all online users.

**Comment by daira** [comment]
Agreed that Pours should not be anchored to very recent blocks. While that reduces the scope of the problem, it doesn't eliminate it: there are other reasons why the blockchain might evolve in a way that prevents a submitted Pour from succeeding.

**Comment by daira** [comment]
Note that section 6.4 of the Zerocash paper describes a somewhat related issue:

> Additionally, although in theory all users have a single view of the block chain, a powerful
attacker could potentially fabricate an additional block solely for a targeted user. Spending any
coins with respect to the updated Merkle tree in this “poison-pill” block will uniquely identify the
targeted user. To mitigate such attacks, users should check with trusted peers their view of the
block chain and, for sensitive transactions, only spend coins relative to blocks further back in the
ledger (since creating the illusion for multiple blocks is far harder).

**Comment by daira** [comment]
I think there's nothing we can do about this information leak. Anyone have any ideas on how a wallet could mitigate it?

**Comment by ebfull** [comment]
I don't think there's anything we can do about this, and I don't really find it concerning anyway. I nominate to close.

**Comment by daira** [comment]
The only things we can think of to mitigate this are:
 * to choose the policy for selecting the anchor, e.g. to not choose a too-recent anchor, which is #93;
 * to offer sufficient fees;
 * to randomize the positions of the input and output coins, which is #778.

The information leakage is only that the Pours spend one or more of the same coins. Also the positions of reused coins could potentially leak information, depending on the wallet strategy for merging coins and making change, unless #778 is done.

**Comment by daira** [comment]
Reopening because this should be documented.

**Comment by daira** [comment]
A possible defence / minimization of this information leakage is that if we have a note that we've failed to send (and therefore have revealed its nullifier), then we only spend that note to ourself in a fully shielded transaction.

**Comment by zookozcash** [comment]
Minor correction: the leakage isn't from "failing to send" a transaction, the leakage is from sending a transaction (i.e. revealing the transaction to someone else, including but not limited to revealing it to miners or others by posting it to the mempool) _when that transaction doesn't get finalized in the immutable blockchain_.

**Comment by zookozcash** [comment]
A consequence of this is that finality is necessary for (this particular detail of) privacy.
